exerciser/src/main.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::Context;
+use log::{info, trace};
+use pulldown_cmark::{CowStr, Event, Parser, Tag};
+use std::{
+    env::args,
+    fs::{create_dir, create_dir_all, read_to_string, File},
+    io::Write,
+    path::Path,
+    process::exit,
+};
+
+const INCLUDE_START: &str = "{{#include ";
+const INCLUDE_END: &str = "}}";
+
+fn main() -> anyhow::Result<()> {
+    pretty_env_logger::init();
+
+    let args = args().collect::<Vec<_>>();
+
+    if args.len() != 3 {
+        eprintln!("Usage:");
+        eprintln!(
+            "  {} <src/exercises/exercise.md> <output directory>",
+            args[0]
+        );
+        exit(1);
+    }
+
+    let input_filename = Path::new(&args[1]);
+    let output_directory = Path::new(&args[2]);
+
+    create_dir(output_directory).with_context(|| {
+        format!("Failed to create output directory {:?}", output_directory)
+    })?;
+
+    let input_directory = input_filename
+        .parent()
+        .with_context(|| "Input file has no parent directory.")?;
+    let input_contents = read_to_string(input_filename)
+        .with_context(|| format!("Failed to open {:?}", input_filename))?;
+    let parser = Parser::new(&input_contents);
+
+    let mut next_filename: Option<String> = None;
+    let mut current_file: Option<File> = None;
+    for event in parser {
```

Comment:
Parsing Markdown like this makes me a little nervous: the coupling between the source and this script is far from obvious. Even reading this now, it's not obvious to me what the Markdown looks like which will trigger some useful action in this code :slightly_smiling_face: 

One thing I've noticed is that your pages don't follow the same style as the rest of the pages. Compare

![image](https://user-images.githubusercontent.com/89623/229169626-9dc09920-d0c5-4c4c-a8c3-97b489798511.png)

with the existing style:

![image](https://user-images.githubusercontent.com/89623/229169735-9cab76e9-51d7-49e5-9924-ad7018f73f61.png)

The way I would like to handle this is with a new preprocessor, similar to the `links` preprocessor which is responsible for resolving the `{{#include` syntax. The new preprocessor, call it `file` would take

~~~markdown
{{#file some/path/foo.rs:Foo}}
~~~

and expand it to

~~~markdown
_some/path/foo.rs_:

```rust
{{#include some/path/foo.rs:Foo}}
```
~~~

(It can guess the syntax based on the file extension.)

That would give us a consistent way to show files in the course.

Ideally, the preprocessor would also be able to emit something which we can recognize clearly in the rendered book (probably in a book rendered as Markdown). I'm thinking it could emit `<!-- Start include -->` or similar and we would then find that in a [`Event::Html`](https://docs.rs/pulldown-cmark/latest/pulldown_cmark/enum.Event.html#variant.Html), if I'm guessing correctly.

I would feel better about that than assigning as special meaning to a line of code followed by a code block.

Reply:
> because it needs to handle including multiple sections interspersed with inline code

Right... I had not thought of that. The special comments are also nice and very flexible :+1: